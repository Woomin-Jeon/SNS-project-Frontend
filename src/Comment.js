/* eslint-disable react/prop-types */
/* eslint-disable react/jsx-one-expression-per-line */
/* eslint-disable react/jsx-wrap-multilines */
/* eslint-disable implicit-arrow-linebreak */
import React from 'react';
import profile from './profile.jpeg';

const Comment = ({ commentState, post }) => {
  const { comment } = commentState;
  const appropriateComment = [];

  for (let i = 0; i < comment.length; i++) {
    if (post.id === comment[i].id) {
      appropriateComment.push(comment[i]);
    }
  }

  return (
    <div className="comment-contents">
      {appropriateComment.map((v) =>
        <div key={v.writer}>
          <span className="comment-main">
            <img className="comment-image" src={profile} alt="" width="3.5%" />
            {v.writer} : {v.statement}
          </span>
          <button type="button" className="comment-thumb">좋아요</button>
          <span className="comment-thumb-count">{v.commentsThumbCount}</span>
        </div>)}
    </div>
  );
};

export default Comment;
