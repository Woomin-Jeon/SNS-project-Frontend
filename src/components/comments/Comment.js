import React, { useState, useEffect } from 'react';
import func from '../../function';
import ChildCommentBox from './ChildCommentBox';

const callAPI = async (commentState, setCommentState, loginState, setLoginState) => {
  const { postComments } = await func.getComments();
  const { userStore } = await func.getUsers();

  setLoginState({ ...loginState, users: [...userStore] });

  setCommentState({
    ...commentState,
    comment: [...postComments.reverse()],
  });
}

function Comment({
  specificPost,
  postState,
  setPostState,
  currentUserState,
  commentState,
  setCommentState,
  loginState,
  setLoginState,
}) {
  const { comment } = commentState;
  const { userName, id } = currentUserState;
  const { uniqueKey } = specificPost;
  const { users } = loginState;
  const [temptState, setTemptState] = useState('');

  useEffect(() => {
    callAPI(commentState, setCommentState, loginState, setLoginState);
  }, []);

  const setCommentTemptStatement = (temptState) => {
    setTemptState(temptState);
  };

  const handleAddComment = async () => {
    if (temptState.trim()) {
      const { postComments } = await func.addComment(uniqueKey, id, userName, temptState);
      const { timeLinePosts } = await func.plusCommentCount(uniqueKey);

      setCommentState({ ...commentState, comment: [...postComments.reverse()] });
      setPostState({ ...postState, post: [...timeLinePosts.reverse()] });
      setTemptState('');
    }
  };

  const handleCommentThumbCount = async (specificComment) => {
    const { postComments } = await func.plusCommentThumbCount(
      specificComment.uniqueKey,
      currentUserState.id,
    );

    setCommentState({ ...commentState, comment: [...postComments.reverse()] });
  };

  const ChildCommentButtonClicked = (specificComment) => {
    setCommentState(func.openChildCommentBox(commentState, specificComment));
  };
  return (
    <>
      <div className="comment">
        <span className="comment-word">댓글달기</span>
        <input
          className="comment-input-box"
          type="text"
          value={temptState}
          onChange={(e) => setCommentTemptStatement(e.target.value)}
        />
        <button
          className="comment-button"
          type="button"
          onClick={() => handleAddComment()}
        >
          입력
        </button>
      </div>
      <div className="comment-contents">
        {comment.filter((v) => specificPost.uniqueKey === v.id).map((v, index) =>
          <div className="comment-each" key={`Comment${index}`}>
            <span>
              <img
                className="comment-profile-image"
                src={users[users.findIndex(user => user.id === v.writerID)].profile}
                alt=""
                width="3.5%"
              />
              <div className="comment-statement">
                <span className="comment-writer">{v.writer}</span>
                <span className="comment-colon">:</span>
                <span className="comment-saying">{v.statement}</span>
              </div>
            </span>
            <button
              className="comment-thumb"
              type="button"
              onClick={() => ChildCommentButtonClicked(v)}
            >
              대댓글
            </button>
            <button
              type="button"
              className="comment-thumb"
              onClick={() => handleCommentThumbCount(v)}
            >
              좋아요
            </button>
            <span className="comment-thumb-count">{v.commentThumbCount.length}</span>
            {v.isChildCommentFunctionOn
              ? (
                <div className="child-comment-main">
                  <ChildCommentBox
                    commentState={commentState}
                    setCommentState={setCommentState}
                    parentsComment={v}
                    currentUserState={currentUserState}
                  />
                </div>
              ) : <></>}
            <div>
              {v.childComment.map((k, index) =>
                <div key={index} className="child-contents">
                  └ {k.name} :<span> </span>{k.statement}
                </div>)}
            </div>
          </div>)}
      </div>
    </>
  );
}

export default Comment;
