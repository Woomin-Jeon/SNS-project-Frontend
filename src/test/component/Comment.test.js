import React from 'react';
import {render} from "enzyme";
import Comment from "../../components/comments/Comment";

describe('Comment', () => {
  let currentUserState;
  let postState;
  let setPostState;
  let commentState;
  let setCommentState;
  let loginState;
  let setLoginState;
  let p;

  beforeEach(() => {
    currentUserState = {
      id: 'Random_ID',
      userName: 'Woomin',
    };
    loginState = {
      users: [ { id: 'Random_ID' }]
    };
    commentState = {
      comment: []
    };
    p = {
      time: [2020, 4, 21, 5, 16],
      thumbCount: [],
    };
  });

  it('renders well', () => {
    const component = render(
      <Comment
        specificPost={p}
        setPostState={setPostState}
        postState={postState}
        currentUserState={currentUserState}
        commentState={commentState}
        setCommentState={setCommentState}
        loginState={loginState}
        setLoginState={setLoginState}
      />
    );

    expect(component).toMatchSnapshot();
    expect(component.find('.comment-word').text()).toBe('댓글달기');
    expect(component.find('.comment-button').text()).toBe('입력');
  });
});
