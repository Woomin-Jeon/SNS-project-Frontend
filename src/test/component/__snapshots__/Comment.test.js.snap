// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Comment renders well 1`] = `ShallowWrapper {}`;
